Fix for finding child output columns when parent is union while join pruning

DevChattopadhyay opened the pull request:
Fix issue  #16932

## Terminology used:

`Derived Output columns:` Output columns defined by each operator

`Output Columns:` The intersection of the operator's derived output columns and
the operator's parent's combined output pred columns. If the operator is the
root operator, its output columns would be the intersection of the operator's
derived output columns and the query's output columns.

`Combined Output Pred Columns:` A set containing operator's output columns and
its predicate's (if present) derived output columns

## Root Cause Analysis
-------------------------

Consider the below tree
```
+--A(Logical)
   |--B (Logical)
   |  |----C (Logical)
   |  |----D (Logical)
   |  +----E (Scalar)
   |
   +--F (Scalar)
```
In order to perform left join pruning in ORCA, one of the step is that each
operator(Logical) determines two sets. One set containing output columns of the
operator and second set containing combined output pred columns: For e.g. in
the above tree, operator A will create two sets. One set containing output
columns of A (let's call this set S1) and second set containing combined output
pred columns (let's call this set S2). In order to determine those two sets the
below algorithm is employed (Suppose we are determining for operator A)
1.) The operator A will find it's derived output columns and will take a
intersection with the query's output. The common values will be inserted into
set S1 and S2. The set S1 now contains the output columns of operator A.
2.)If a scalar operator is present for operator A then the derived output
columns of the scalar operator are also inserted into the set S2 (In this case
operator F's derived output columns)
3.) Now the set S2 will contain combined output pred columns for operator A.
4.) This set S2 is passed to the next function call in which the output columns
of operator B will be determined.

Now consider the below query:

**Setup:**
```
create table t1(a int primary key, b int) ;
create table t2(a int primary key, b int) ;
```

**Query:**

```
explain select t2.b from t1 left join t2 on t1.a = t2.a union all select t2.b
from t1 left join t2 on t1.a = t2.a;
```

**Algebrized tree:**

```
+--CLogicalUnionAll Output: ("b" (10)), Input: [("b" (10)), ("b" (28))]
   |--CLogicalLeftOuterJoin
   |  |--CLogicalGet "t1" ("t1"), Columns: ["a" (0), "b" (1)]
   |  |--CLogicalGet "t2" ("t2"), Columns: ["a" (9), "b" (10)]
   |  +--CScalarCmp (=)
   |     |--CScalarIdent "a" (0)
   |     +--CScalarIdent "a" (9)
   +--CLogicalLeftOuterJoin
      |--CLogicalGet "t1" ("t1"), Columns: ["a" (18), "b" (19)]
      |--CLogicalGet "t2" ("t2"), Columns: ["a" (27), "b" (28)]
      +--CScalarCmp (=)
         |--CScalarIdent "a" (18)
         +--CScalarIdent "a" (27)
```

Based on the above algorithm mentioned
The derived output columns of CLogicalSetOp
(CLogicalUnion/UnionAll/Intersect/IntersectAll/Except/ExceptAll) are from its
outer child. Based on the algorithm mentioned above, CLogicalUnionAll will
create two sets. One set containing output columns of itself and second set
containing combined output pred columns . The first set will contain b(10) and
the second set will also contain b(10). This second set will be passed to the
outer and inner CLogicalLeftOuterJoin operator for determining there output
columns. The outer CLogicalLeftOuterJoin will find its derived output columns
and will intersect it with the set passed by CLogicalUnionAll containing b(10).
After intersection we will get b(10) as the output column for outer
CLogicalLeftOuterJoin.  Similarly the inner CLogicalLeftOuterJoin will find its
derived output columns and will intersect it with the set passed by
CLogicalUnionAll containing b(10). But here the intersection will lead to a
empty output column set for inner CLogicalLeftOuterJoin. Now for pruning a left
join in a query, one of the condition that ORCA checks is that all output
columns of CLogicalLeftOuterJoin should only belong from the outer relation.
The method performing this check doesn't handle the case when the output
columns set is empty. Due to this all the checks pass in that method and ORCA
proceeds to prune the Join. So basically the issue is that ORCA is not able to
properly determine the output columns for the child operators when the parent
is CLogicalSetOp which is leading to unwanted pruning of joins. A similar issue
was fixed in PR #16690 which ensured that the combined output pred columns of
CLogicalSetOp is not empty because an empty set was leading to unwanted pruning
of left join. But that fix was not complete. It was incorrectly determining
combined output pred columns  of CLogicalSetOp which lead to #16932.

## Fix
--------
Updating the logic for determining the combined output pred columns  of
CLogicalSetOp. Since in the case of CLogicalSetOp, the output columns are from
the outer child, the inner child will always have empty output columns which is
not completely correct. So we need to also include the inner child columns in
the combined output pred columns  for CLogicalSetOP.

dgkimura commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -418,6 +426,59 @@ CLeftJoinPruningPreprocessor::ComputeOutputColumns(
 	// Computing output columns of the child tree
 	childs_output_columns->Include(output_columns);
+
+	// Consider the below algebrized tree
+	// CLogicalUnionAll Output: ("b" (10)), Input: [("b" (10)), ("b" (28))]
+	//  |--CLogicalLeftOuterJoin
+	//  |  |--CLogicalGet "t1" ("t1"), Columns: ["a" (0), "b" (1)]
+	//  |  |--CLogicalGet "t2" ("t2"), Columns: ["a" (9), "b" (10)]
+	//  |  +--CScalarCmp (=)
+	//  |     |--CScalarIdent "a" (0)
+	//  |     +--CScalarIdent "a" (9)
+	//  +--CLogicalLeftOuterJoin
+	//     |--CLogicalGet "t1" ("t1"), Columns: ["a" (18), "b" (19)]
+	//     |--CLogicalGet "t2" ("t2"), Columns: ["a" (27), "b" (28)]
+	//     +--CScalarCmp (=)
+	//        |--CScalarIdent "a" (18)
+	//        +--CScalarIdent "a" (27)
+	// In this case the output_columns of CLogicalUnionAll will only contain
+	// b(10). In the case of CLogicalUnion or CLogicalUnionAll , the output
+	// columns are from the outer child. Due to this the output_columns of inner
+	// CLogicalLeftOuterJoin will be empty. So join pruning will not be possible
+	// in such cases. The Input in CLogicalUnion/CLogicalUnionAll signifies the
+	// columns from the childs. So including those columns in the
+	// childs_output_columns of CLogicalUnionAll to perform pruning of the inner
+	// child if possible.
```
IIUC, the issue is that we pruned the child that projected "b" (28) which is needed as input for CLogicalUnionAll.  As an alternative solution, is it possible to update CLogicalUnionAll to take as input "b" (19) instead of "b" (28) and continue to allow the join to be pruned?

DevChattopadhyay replied:
Query:
```
explain select t2.b from t1 left join t2 on t1.a = t2.a union all select t2.b
from t1 left join t2 on t1.a = t2.a;
```

@dgkimura the issue is actually that we pruned the inner CLogicalLeftOuterJoin of union. If you observe the inner side query of union (select t2.b from t1 left join t2 on t1.a = t2.a) we are selecting the column t2.b . Join pruning should not happen in the first place if we select a column from inner relation t2. We should not allow the join to be pruned.

dgkimura replied:
Ah, I see.  Thanks for that clarification.

khannaekta commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -418,6 +426,59 @@ CLeftJoinPruningPreprocessor::ComputeOutputColumns(
+	// childs_output_columns of CLogicalUnionAll to perform pruning of the inner
+	// child if possible.
+
+	BOOL isLogicalUnion = (pexpr->Pop()->Eopid() == COperator::EopLogicalUnion);
+	BOOL isLogicalUnionAll =
+		(pexpr->Pop()->Eopid() == COperator::EopLogicalUnionAll);
+
+	if (isLogicalUnion || isLogicalUnionAll)
+	{
+		CLogicalUnion *pop = nullptr;
+
+		if (isLogicalUnion)
+		{
+			pop = CLogicalUnion::PopConvert(pexpr->Pop());
+		}
+		else
+		{
+			pop = CLogicalUnionAll::PopConvert(pexpr->Pop());
+		}
+
+		for (ULONG uli = 0; uli < pop->PdrgpdrgpcrInput()->Size(); uli++)
+		{
+			for (ULONG ulj = 0; ulj < (*pop->PdrgpdrgpcrInput())[uli]->Size();
+				 ulj++)
+			{
+				childs_output_columns->Include(
+					(*(*pop->PdrgpdrgpcrInput())[uli])[ulj]);
+			}
+		}
+	}
```
If we have this special logic for Union/UnionALL, what about the case for other logical operators like `CLogicalDifference`/ALL and `CLogicalIntersect`/ALL? Shouldn't we see the same issue?

DevChattopadhyay replied:
Thanks @khannaekta for pointing that out. I completely missed that.  We will actually observe the same issue for these logical ops also. I have updated the code to handle `CLogicalDifference/ALL` and `CLogicalIntersect/ALL`

khannaekta commented:
Regression on this PR branch. For the following queries we miss on join pruning, which we currently do on main branch:
```
create table t1(a int primary key, b int) ;
create table t2(a int primary key, b int) ;

explain select t2.b from t1 left join t2 on t1.a = t2.a
except
select t1.b from t1 left join t2 on t1.a = t2.a;

explain select t2.b from t1 left join t2 on t1.a = t2.a 
intersect
select t1.b from t1 left join t2 on t1.a = t2.a;
```

DevChattopadhyay commented:
> Regression on this PR branch. For the following queries we miss on join pruning, which we currently do on main branch:
> 
> ```
> create table t1(a int primary key, b int) ;
> create table t2(a int primary key, b int) ;
> 
> explain select t2.b from t1 left join t2 on t1.a = t2.a
> except
> select t1.b from t1 left join t2 on t1.a = t2.a;
> 
> explain select t2.b from t1 left join t2 on t1.a = t2.a 
> intersect
> select t1.b from t1 left join t2 on t1.a = t2.a;
> ```

I have updated the code.. Now this PR branch will be able to prune the joins.

gpopt commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -418,6 +430,89 @@ CLeftJoinPruningPreprocessor::ComputeOutputColumns(
+		case COperator::EopLogicalIntersect:
+		{
+			pop = CLogicalIntersect::PopConvert(pexpr->Pop());
+			isLogicalSetOp = true;
+			break;
+		}
+		case COperator::EopLogicalIntersectAll:
+		{
+			pop = CLogicalIntersectAll::PopConvert(pexpr->Pop());
+			isLogicalSetOp = true;
+			break;
+		}
+		case COperator::EopLogicalDifference:
+		{
+			pop = CLogicalDifference::PopConvert(pexpr->Pop());
+			isLogicalSetOp = true;
+			break;
+		}
+		case COperator::EopLogicalDifferenceAll:
+		{
+			pop = CLogicalDifferenceAll::PopConvert(pexpr->Pop());
+			isLogicalSetOp = true;
+			break;
+		}
+		default:
+			break;
+	}
+
+	if (isLogicalSetOp)
+	{
+		for (ULONG uli = 0; uli < pop->PdrgpdrgpcrInput()->Size(); uli++)
+		{
+			for (ULONG ulj = 0; ulj < (*pop->PdrgpdrgpcrInput())[uli]->Size();
+				 ulj++)
+			{
+				childs_output_columns->Include(
+					(*(*pop->PdrgpdrgpcrInput())[uli])[ulj]);
+			}
+		}
+	}
```
Why not this?
```cpp
if (CUtils::FLogicalSetOp(pexpr->Pop()))
{
    CLogicalSetOp *pop = CLogicalSetOp::PopConvert(pexpr->Pop());
    for (ULONG i = 0; i < pop->PdrgpdrgpcrInput()->Size(); i++)
		{
            // please us a temp var to save (*pop->PdrgpdrgpcrInput())[i]
			for (ULONG j = 0; j < (*pop->PdrgpdrgpcrInput())[i]->Size();
				 j++)
			{
				childs_output_columns->Include(
					(*(*pop->PdrgpdrgpcrInput())[i])[j]);
			}
		}
}
```

Recommend using `i`, `j`, instead of `uli`, `ulj`, which are counter intuitive.

DevChattopadhyay replied:
Updated the code.

dgkimura commented on src/test/regress/sql/join_gp.sql:
```diff
@@ -838,9 +838,51 @@ drop table barJoinPruning;
 --
 create table foo(a int primary key, b int);
 create table bar(a int primary key, b int);
+
 explain  select foo.a, bar.b from foo left join bar on foo.a = bar.a
  union
    select foo.a, bar.b from foo join bar on foo.a = bar.a;
+
+-- join will be pruned
+explain select foo.a,foo.b from foo left join bar on foo.a=bar.a union
+select bar.a,bar.b from bar left join foo on foo.a=bar.a;
```
Is it worth adding a few rows of data and some NULL values, and then checking the results?

I feel that looking at a bad plan is not always immediately obvious and may evade quick eye-checks, but wrong results are less likely to go unseen.

DevChattopadhyay replied:
@dgkimura I have updated the test cases. Please take a second look.

dgkimura reviewed (approved):
Looks good.  I'd suggest to also check the results that the queries return in the tests.

THANATOSLAVA commented:
What's the relation between this PR and [16690](https://github.com/greenplum-db/gpdb/pull/16690)? The PR descriptions are strikingly similar. Did 16690 take the wrong approach?

DevChattopadhyay commented:
> What's the relation between this PR and [16690](https://github.com/greenplum-db/gpdb/pull/16690)? The PR descriptions are strikingly similar. Did 16690 take the wrong approach?

@THANATOSLAVA 
The PR #16960 and #16690 on a high level are solving the exact same issue i.e. calculating the childs_output_columns when the operator is CLogicalUnion/CLogicalIntersect/CLogicalExcept/*ALL 
The previous PR's (#16690) approach was not completely wrong but was incomplete. That PR only fixes the cases that we don't receive empty pcrsOutput for the child operators of CLogicalUnion and other SetOp. But it was determining the childs_output_columns of CLogicalUnion incorrectly.
This PR (#16960) will handle all the cases (will not generate empty pcrsOutput columns for children of SetOp and will determine the  childs_output_columns for SetOp correctly).

khannaekta commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -12,6 +12,11 @@
 #include "gpopt/operators/CLeftJoinPruningPreprocessor.h"
 
 #include "gpopt/base/CColRefSetIter.h"
+#include "gpopt/operators/CLogicalDifference.h"
+#include "gpopt/operators/CLogicalDifferenceAll.h"
+#include "gpopt/operators/CLogicalIntersect.h"
+#include "gpopt/operators/CLogicalIntersectAll.h"
+#include "gpopt/operators/CLogicalUnionAll.h"
```
We probably don't need these includes anymore

DevChattopadhyay replied:
Removed

THANATOSLAVA commented:
> @THANATOSLAVA The PR #16960 and #16690 on a high level are solving the exact same issue i.e. calculating the childs_output_columns when the operator is CLogicalUnion/CLogicalIntersect/CLogicalExcept/*ALL The previous PR's (#16690) approach was not completely wrong but was incomplete. That PR only fixes the cases that we don't receive empty pcrsOutput for the child operators of CLogicalUnion and other SetOp. But it was determining the childs_output_columns of CLogicalUnion incorrectly. This PR (#16960) will handle all the cases (will not generate empty pcrsOutput columns for children of SetOp and will determine the childs_output_columns for SetOp correctly).

This is a bit clearer now. From what I gather, this PR corrects the output columns of the second and subsequent children of the union all. Prior to this PR, the incorrectly derived output columns were utilized to wrongly prune the second and subsequent LOJ children of the union all?

Can we do the following for the PR description?

1. Remove all references to variable names. Variable names are contextual. Readers need to look into the function to understand the naming nuances. Let's try to describe `childs_output_columns` and `pcrsOutput` using plain English. Also we should be consistent with the usage of "current" (node) in the context of a tree. If the maximum hierarchy that we examine here is 2, referring to nodes as "child" and "parent" would cause less ambiguity.
2. Reference PR #16690, its contribution and flaw. This helps reader justify the value of this PR.

DevChattopadhyay commented:
> > @THANATOSLAVA The PR #16960 and #16690 on a high level are solving the exact same issue i.e. calculating the childs_output_columns when the operator is CLogicalUnion/CLogicalIntersect/CLogicalExcept/*ALL The previous PR's (#16690) approach was not completely wrong but was incomplete. That PR only fixes the cases that we don't receive empty pcrsOutput for the child operators of CLogicalUnion and other SetOp. But it was determining the childs_output_columns of CLogicalUnion incorrectly. This PR (#16960) will handle all the cases (will not generate empty pcrsOutput columns for children of SetOp and will determine the childs_output_columns for SetOp correctly).
> 
> This is a bit clearer now. From what I gather, this PR corrects the output columns of the second and subsequent children of the union all. Prior to this PR, the incorrectly derived output columns were utilized to wrongly prune the second and subsequent LOJ children of the union all?
> 
> Can we do the following for the PR description?
> 
> 1. Remove all references to variable names. Variable names are contextual. Readers need to look into the function to understand the naming nuances. Let's try to describe `childs_output_columns` and `pcrsOutput` using plain English. Also we should be consistent with the usage of "current" (node) in the context of a tree. If the maximum hierarchy that we examine here is 2, referring to nodes as "child" and "parent" would cause less ambiguity.
> 2. Reference PR [Fix LeftJoinPruning pruns essential left join #16690](https://github.com/greenplum-db/gpdb/pull/16690), its contribution and flaw. This helps reader justify the value of this PR.

Updated the commit message. Let me know if any modifications are required.

THANATOSLAVA commented:
> The operator A determines a set containing columns which are being output
by all its child operators combined and pass that information to the next
function call in which the columns being output by operator B will be
determined.

Here's my interpretation of the above. We want to derive `B`'s output columns. In order to do that, we need to obtain the output columns of all the `A`'s children's, which is `B` in this case.

So we derive `B`'s output columns using `B`'s output columns?

THANATOSLAVA commented:
> In order to perform left join pruning in ORCA, one of the step is that each
operator(Logical) determines two sets. One set containing columns which are
being output by itself and second set containing columns which are being output
by all its child operators(Logical) combined.

Can we add a statement here how the two column sets are used for left join pruning?

DevChattopadhyay commented:
> > The operator A determines a set containing columns which are being output
> > by all its child operators combined and pass that information to the next
> > function call in which the columns being output by operator B will be
> > determined.
> 
> Here's my interpretation of the above. We want to derive `B`'s output columns. In order to do that, we need to obtain the output columns of all the `A`'s children's, which is `B` in this case.
> 
> So we derive `B`'s output columns using `B`'s output columns?


Lets take an example in which the algebrized tree is similar to the one used in commit message

**Setup:**
```
create table foo (a int primary key,b int, c int);
create table bar ( p int primary key, q int, r int);
```

**Query:**
```
explain select foo.c from foo left join bar on foo.a=bar.p where foo.b>10;
```

**Algebrized Tree:**
```
+--CLogicalSelect
   |--CLogicalLeftOuterJoin
   |  |--CLogicalGet "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Key sets: {[0], [3,9]}
   |  |--CLogicalGet "bar" ("bar"), Columns: ["p" (10), "q" (11), "r" (12), "ctid" (13), "xmin" (14), "cmin" (15), "xmax" (16), "cmax" (17), "tableoid" (18), "gp_segment_id" (19)] Key sets: {[0], [3,9]}
   |  +--CScalarCmp (=)
   |     |--CScalarIdent "a" (0)
   |     +--CScalarIdent "p" (10)
   +--CScalarCmp (>)
      |--CScalarIdent "b" (1)
      +--CScalarConst (10)
```

Below table gives the view of all the variables during function calls

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/cdev/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/cdev/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:16.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Menlo, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{text-align:center;
	vertical-align:middle;
	border:.5pt solid windowtext;}
.xl66
	{text-align:center;
	vertical-align:middle;
	border:.5pt solid windowtext;
	white-space:normal;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Operator | derived output   columns | pcrsOutput(the   columns being output) by the operator | output columns | childs_output_columns
-- | -- | -- | -- | --
CLogicalSelect | c(2) | c(2) | c(2) | b(1),c(2)
CLogicalLeftOuterJoin | "a"   (0), "b" (1), "c" (2), "ctid" (3),   "xmin" (4), "cmin" (5), "xmax" (6),   "cmax" (7), "tableoid" (8), "gp_segment_id"   (9), "p" (10), "ctid" (13), "xmin" (14),   "cmin" (15), "xmax" (16), "cmax" (17),   "tableoid" (18), "gp_segment_id" (19)" | b(1),c(2) | b(1),c(2) | "a"   (0), "b" (1), "c" (2), "p" (10)"



</body>

</html>


So for CLogicalLeftOuterJoin we see that the derived output columns is a set containing columns `"a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9), "p" (10), "ctid" (13), "xmin" (14), "cmin" (15), "xmax" (16), "cmax" (17), "tableoid" (18), "gp_segment_id" (19)"`. But the set created by CLogicalSelect which contains the columns being output by all child operator (pcrsOutput for CLogicalLeftOuterJoin)  contains `b(1),c(2)`. We take the intersection off these two sets and determine the ouptut columns of CLogicalLeftOuterJoin which comes out to be `b(1),c(2)`

DevChattopadhyay commented:
> > In order to perform left join pruning in ORCA, one of the step is that each
> > operator(Logical) determines two sets. One set containing columns which are
> > being output by itself and second set containing columns which are being output
> > by all its child operators(Logical) combined.
> 
> Can we add a statement here how the two column sets are used for left join pruning?

Updated the commit message

THANATOSLAVA commented:
> Operator	derived output columns	pcrsOutput(the columns being output) by the operator

You would probably have guessed my next question -- what's the difference between "derived output columns" and "columns being output" (without getting into the code details)

DevChattopadhyay commented:
> You would probably have guessed my next question -- what's the difference between "derived output columns" and "columns being output" (without getting into the code details)

@THANATOSLAVA I have updated the commit message with the terminologies.Please do take a look.

gpopt commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -12,6 +12,7 @@
 #include "gpopt/operators/CLeftJoinPruningPreprocessor.h"
 
 #include "gpopt/base/CColRefSetIter.h"
+#include "gpopt/operators/CLogicalDifference.h"
```
Is this needed?

DevChattopadhyay replied:
Yes not required..We need to include CLogicalSetOp.h

gpopt reviewed (approved):
+1, after the comment is addressed.

THANATOSLAVA commented:
> Query's output: The resulting columns of a query.

Can we get rid of this one?

> Derived Output columns: The derived output columns of an operator is basically the result set of that operator.
Output Columns: Output columns signifies the columns that are being selected from the result set of an operator(Logical) to produce the query's output.

Let's make it really plain. Feel free to touch up but my understanding is

**Derived output columns:** All the columns referenced by a query plus all the system columns
**Output columns:** The columns a query returns

THANATOSLAVA commented:
> 2.) If a scalar operator is present for operator A then the derived output columns of the scalar operator are also inserted into the set S2.

I assume you meant adding operator F's derived output columns to S2? Can we add that so we make use of your illustration.

THANATOSLAVA commented:
I'm having trouble wrapping my head around this

> Combined Child Output columns: A group of output columns of all the child operators(Logical) of an operator.
...
1.) The operator A will find it's derived output columns and will take a intersection with the query's output. The common values will be inserted into set S1 and S2. The set S1 now contains the output columns of operator A.
2.) If a scalar operator is present for operator A then the derived output columns of the scalar operator are also inserted into the set S2.
3.) Now the set S2 will contain combined child output columns for operator A.
4.) This set S2 is passed to the next function call in which the output columns of operator B will be determined.

Are you trying to say operator's A's "combined child output columns" (S2) includes operator's A's output columns and operator F's derived output columns?

THANATOSLAVA commented:
> Based on the above algorithm mentioned
The derived output columns of CLogicalSetOp (CLogicalUnion/UnionAll/Intersect/IntersectAll/Except/ExceptAll) are from its outer child. CLogicalUnionAll will create two sets.

--> 


The derived output columns of CLogicalSetOp (CLogicalUnion/UnionAll/Intersect/IntersectAll/Except/ExceptAll) are from its outer child. **Based on the algorithm mentioned above**, CLogicalUnionAll will create two sets.

THANATOSLAVA commented:
If I am understanding correctly, the issue addressed by this PR pertains to S2 only. If so, we should leave S1 out of the description.

DevChattopadhyay commented:
> > Derived Output columns: The derived output columns of an operator is basically the result set of that operator.
> > Output Columns: Output columns signifies the columns that are being selected from the result set of an operator(Logical) to produce the query's output.
> 
> Let's make it really plain. Feel free to touch up but my understanding is
> 
> The following concept pertains an operator **Derived output columns:** output columns defined by each operator **Output columns:** The intersection between the operator's derived output columns and the query's output columns

Done

DevChattopadhyay commented:
> ur illustration.

Yes i meant adding operator F's derived output columns to S2. Added this in the commit message.

DevChattopadhyay commented:
> operator's A's "combined child output columns" (S2) includes operator's A's output columns and operator F's derived output columns

Yes i meant operator's A's "combined child output columns" (S2) includes operator's A's output columns and operator F's derived output columns. Renamed the variable to `combined_output_pred_columns` and also modified the commit message.

DevChattopadhyay commented:
> > Based on the above algorithm mentioned
> > The derived output columns of CLogicalSetOp (CLogicalUnion/UnionAll/Intersect/IntersectAll/Except/ExceptAll) are from its outer child. CLogicalUnionAll will create two sets.
> 
> -->
> 
> The derived output columns of CLogicalSetOp (CLogicalUnion/UnionAll/Intersect/IntersectAll/Except/ExceptAll) are from its outer child. **Based on the algorithm mentioned above**, CLogicalUnionAll will create two sets.

Updated the commit message

DevChattopadhyay commented:
> If I am understanding correctly, the issue addressed by this PR pertains to S2 only. If so, we should leave S1 out of the description.

The issue pertains to both S1 (output columns of an operator) and S2 (combined_output_pred_columns). We were actually determining S2 wrong which is used to determine S1 for the child operators. So i think both S1 and S2 are relevant for this issue.

THANATOSLAVA commented:
> > If I am understanding correctly, the issue addressed by this PR pertains to S2 only. If so, we should leave S1 out of the description.
> 
> The issue pertains to both S1 (output columns of an operator) and S2 (combined_output_pred_columns). We were actually determining S2 wrong which is used to determine S1 for the child operators. So i think both S1 and S2 are relevant for this issue.

I see. So an operator's S1 is the intersection of its derived output columns and its parent's S2 (which is the parent's S1 plus the parent's predicate columns if any). In that case, we have to update our definition accordingly --

`Output Columns` (S1): The intersection of the operator's `derived output columns` and the operator's parent's `combined output pred columns`. If the operator is the root operator, its `output columns` would be the intersection of the operator's `derived output columns` and the query's output columns.

Lemme know if you think this makes sense.

DevChattopadhyay commented:
> > > If I am understanding correctly, the issue addressed by this PR pertains to S2 only. If so, we should leave S1 out of the description.
> > 
> > 
> > The issue pertains to both S1 (output columns of an operator) and S2 (combined_output_pred_columns). We were actually determining S2 wrong which is used to determine S1 for the child operators. So i think both S1 and S2 are relevant for this issue.
> 
> I see. So an operator's S1 is the intersection of its derived output columns and its parent's S2 (which is the parent's S1 plus the parent's predicate columns if any). In that case, we have to update our definition accordingly --
> 
> `Output Columns` (S1): The intersection of the operator's `derived output columns` and the operator's parent's `combined output pred columns`. If the operator is the root operator, its `output columns` would be the intersection of the operator's `derived output columns` and the query's output columns.
> 
> Lemme know if you think this makes sense.

Thanks @THANATOSLAVA . This makes more sense. Updated the commit message.

THANATOSLAVA commented on src/backend/gporca/libgpopt/src/operators/CLeftJoinPruningPreprocessor.cpp:
```diff
@@ -417,7 +425,46 @@ CLeftJoinPruningPreprocessor::ComputeOutputColumns(
+	if (CUtils::FLogicalSetOp(pexpr->Pop()))
+	{
+		CLogicalSetOp *pop = CLogicalSetOp::PopConvert(pexpr->Pop());
+		for (ULONG i = 0; i < pop->PdrgpdrgpcrInput()->Size(); i++)
```
Shouldn't we start from index 1, since the 0th input column array is the same as the output column array?

DevChattopadhyay replied:
Nice catch..Yes it should start from index 1. Thanks

THANATOSLAVA commented on src/test/regress/sql/join_gp.sql:
```diff
@@ -836,11 +836,91 @@ drop table barJoinPruning;
+union select foo.a, bar.b from foo join bar on foo.a = bar.a;
+
+-- join will be pruned
+explain (costs off) select foo.b,foo.c from foo left join bar on foo.a=bar.a union
```
Selfish request: can we add a comment for each set of tests -- what the query is about and why the pruning is or isn't expected? I realize I need diff tools to discern differences in between :P

DevChattopadhyay replied:
Updated the testcases with comments.
